### ISSUE TITLE
Attributes don't link to properties when set as a string.

### ISSUE THREAD

**Issue by treshugart** [skatejs/dom-diff#31]
For examle, when you set the `value` attribute of an `<select>` element, the value doesn't propagate to the property thus it doesn't update the selected `<option>`.

**Comment by developit** [comment]
AFAIK the only way to fix this is to switch to setting element properties, which leads to issues with properties that cannot be set (anything on an SVG element, "type" for inputs in IE, etc).

Apologies for the drive-by comment :P

**Comment by treshugart** [comment]
Drive-by's are always welcome (so long as it's constructive) and thanks for the input. I'm unsure as to how to solve this without introducing special naming conventions, which may not be all bad.

Do you have any suggestions?

I've discussed this with a few people and something like the following seems a good starting point.

JSX:

```js
<div attribute="" .property="" onEvent={handler} />
```

JS:

skateDomDiff.vdom.element('div', {
  attribute: '',
  '.property': '',
  onEvent: handler
});

*The `onEvent` syntax already works, but this is just to illustrate all the conventions in one spot. This issue is for dealing with props vs attrs.*

**Comment by developit** [comment]
While it seems nice to have the clear separation, its pretty difficult to know which things are best as attributes and which as properties. Your original issue with the `<select>` value (you'll find the same issue with `<textarea>` when dirty) is a pretty good example of a confusing case. Just my 2¢.

**Comment by treshugart** [comment]
> While it seems nice to have the clear separation, its pretty difficult to know which things are best as attributes and which as properties.

This is kinda why I think it's better to put the onus on the consumer rather than trying to be "too smart". That said, the JSX parser doesn't like a leading `.`, even though it's a valid attribute character.

**Comment by developit** [comment]
Or always use properties. I had to switch from attributes to properties for the same reason:
https://github.com/developit/preact/commit/b84ad6f16d6f8b79025779d491214d5a3492028d

On the plus side, setting properties is often orders of magnitude faster than `setAttribute()`.

**Comment by treshugart** [comment]
> Or always use properties.

But you mentioned above this doesn't work for SVG and `<input type>`, right? How did you solve those scenarios? Maybe it's best to default to setting properties and allow a way to set attributes?

> On the plus side, setting properties is often orders of magnitude faster than setAttribute().

Yeah, IIRC from a React issue their tests were about 3 times faster.

**Comment by treshugart** [comment]
Actually from the looks of that commit, that might work for this use-case as well. Thanks for that.

**Comment by developit** [comment]
SVG is a special-case already, so I , because the root element (and thus the children) must be created using a namespace - that's already an issue of its own, and as such I took the approach of making it an extension/addon library: [preact-svg](git.io/preact-svg).

In Preact at least, I've switched to properties for everything except `"type"`. I know that seems [not great] since type for non-inputs should be mutable, but I already had the fallback in-place for non-reflected properties (ie, anything that does pass the `'name' in node` test.  I was comfortable making that change because of the performance - in my head I can pretend that was the primary reason I did it :P

One side note though:  the original reason I switched from attributes to properties was to mitigate script injection, something I think you'd also be interested in. Here's a simple demo:
https://jsfiddle.net/developit/8vunfq5x/
